llpc/docs/DdnGraphicsRegisterHandler.md

```diff
@@ -0,0 +1,264 @@
+# Graphics Register Handler
+
+# Introduction
+In previous implementation of algorithms, there may include heavy interaction on SRD(Shader readable data), which leads the result of passes to be hardly understanding and introduce redundent operations either.
```

Comment:
I understood this document like this:
*  There's code that wants to generate accesses to graphics registers.
*  Because these operations are low level, they are difficult to maintain and compose poorly, generating redundant operations.
*  To make accessing graphics registers more manageable and generate better code, you introduce a framework that exposes some abstraction over these registers and the operations that can be performed.
*  Because the number of possible operations is restricted, you can compose them together and elide redundant instructions.

Is my understanding correct, or did I get lost somewhere?

Reply:
> Do you mean something along these lines:
> There are many passes that interact with SRD (Shader readable data). To make such code more understandable and less bug-prone, we expose a high-level abstraction (or interface) for accessing SRD.
> (?)
> 
> How does SRD relate to graphics registers?

Sorry, i didn't present it correctly, SRD just told us the value we should use to set related registers. We only need to read it once, modify the value and set them to the target register.
There's no need to write back to SRD again.

Reply:
> * There's code that wants to generate accesses to graphics registers.
> ...
> 
> Is my understanding correct, or did I get lost somewhere?

Yes, but for these registers, the framework didn't touch them really, but generate(maintain) correct values for registers as listed in drivers/llpc/lgc/imported/chip/gfx9/gfx9_plus_merged_registers.h.

Reply:
> Sorry, i didn't present it correctly, SRD just told us the value we should use to set related registers. We only need to read it once, modify the value and set them to the target register.
There's no need to write back to SRD again.

> Sorry, i didn't present it correctly, SRD just told us the value we should use to set related registers. We only need to read it once, modify the value and set them to the target register.
There's no need to write back to SRD again.

Fantastic, this was a piece of information I was missing the whole time. I think it's really important to say that you don't actually modify/write to these registers, but only read them once and later only maintain expression that represent the modified values.